### ISSUE TITLE
[Bug]: silent miscompile — tl.sum(acc, 0) over 3-D tile [RBLOCK,X1,X0] reduces the wrong memory axis on 910B (leading-axis stride swap); same math in 2-D exact; reshape-before-sum not fixed

### ISSUE THREAD

**Issue by gavinwang269** [triton-lang/triton-ascend#2291]
### Summary

A multi-dimensional tile reduction whose **reduction axis is the leading dimension** deterministically produces **wrong runtime values** on **Ascend 910B** — it is a silent miscompile (correct IR, wrong numbers), not a compile failure.

Specifically, `tl.sum(acc, 0)` over a 3-D tile `[RBLOCK, X1, X0]` returns values consistent with the reduction having consumed the **wrong memory axis**: the stride mapping of the reduction axis and an inner `x` axis is swapped/flattened. The error is op- and dtype-independent.

Control: the **exact same offset math** written as a **2-D `[RBLOCK, XBLOCK]`** tile with `tl.sum(acc, 0)` is **numerically exact (maxdiff 0.0)**. Reshaping the 3-D `acc` to 2-D *immediately before* the sum (`tl.reshape(acc, [RBLOCK, X1*X0])` then `tl.sum`) **does NOT** fix it — the lowering is already wrong upstream of the reshape; the index computation itself must be 2-D.

This 3-D leading-axis reduction is the shape that torch inductor's NPU `triton_experimental` backend emits for **broadcast-input backward reductions**, so it manifests there as 33 opinfo rows of large-value gradient errors (abs diff up to 150+; even `aten.add`'s broadcast gradient computes wrong) — but the bug is at the compiler layer and reproduces in pure Triton without inductor.

### Minimal complete example

Save as `repro_3dsum_axis0.py` and run on an idle 910B NPU:

```bash
python repro_3dsum_axis0.py
```

```python
import torch, triton, triton.language as tl

@triton.jit
def k3d(g_ptr, o_ptr, R0, RBLOCK: tl.constexpr, X1: tl.constexpr, X0: tl.constexpr):
    x0 = tl.arange(0, X0)[None, None, :]           # [1,1,5]
    x1 = tl.arange(0, X1)[None, :, None]           # [1,10,1]
    acc = tl.full([RBLOCK, X1, X0], 0, tl.float32)
    for off in tl.range(0, R0, RBLOCK):
        r = off + tl.arange(0, RBLOCK)[:, None, None]   # [16,1,1]
        m = r < R0
        g = tl.load(g_ptr + (x0 + 5 * x1 + 50 * r), m, other=0.0)
        acc = acc + tl.broadcast_to(g, [RBLOCK, X1, X0])
    s = tl.sum(acc, 0)[None, :]                    # reduction over leading axis
    tl.store(o_ptr + (x0 + 5 * x1), s)

@triton.jit
def k2d(g_ptr, o_ptr, R0, XBLOCK: tl.constexpr, RBLOCK: tl.constexpr):
    x = tl.arange(0, XBLOCK)[None, :]
    i, j = x // 5, x % 5                           # same offset math, 2-D flattened
    acc = tl.zeros([RBLOCK, XBLOCK], tl.float32)
    for off in tl.range(0, R0, RBLOCK):
        r = (off + tl.arange(0, RBLOCK))[:, None]
        acc += tl.load(g_ptr + (j + 5 * r + 50 * i), r < R0, other=0.0)
    tl.store(o_ptr + x, tl.sum(acc, 0))

g = torch.arange(500., device='npu').reshape(10, 10, 5) + 1.0
ref = g.sum(dim=1).reshape(50)                     # sum over the stride-5 (X1) axis
o3 = torch.zeros(50, device='npu'); k3d[(1,)](g, o3, 10, RBLOCK=16, X1=10, X0=5, num_warps=1)
o2 = torch.zeros(50, device='npu'); k2d[(1,)](g, o2, 10, XBLOCK=64, RBLOCK=16, num_warps=1)
print("3D maxdiff:", float((o3 - ref).abs().max()))
print("2D maxdiff:", float((o2 - ref).abs().max()))
# Observed (run twice, bit-identical both runs):
#   3D maxdiff: 2025.0   -> o3[k] = sum_r g[k + 50r]  (consumed the stride-50 axis instead)
#   2D maxdiff: 0.0      -> exact
```

### Expected

`tl.sum(acc, axis)` should reduce along the requested logical axis with results matching element-wise semantics (as on the NVIDIA/CPU Triton reference). For a leading-axis 3-D sum, the lowering must not swap/flatten the reduction axis's stride with an inner `x` axis's stride.

### Environment

| Component | Version |
|---|---|
| Hardware | Ascend 910B4 |
| CANN | 9.1.0 |
| triton-ascend | 3.6.0 |
| torch / torch_npu (downstream control) | 2.14.0a0+git / 2.14.0+gitb4c7212 |

### Notes

- Related but distinct: #2247 (A5 2-D reduction lane-drop, `K_POW2 < 8`, wrong-total but stride-correct), #2239 (A5 compile-time IR error on `axis=1` after dynamic-offset masked load). This one is a **910B runtime stride-swap miscompile of a leading-axis 3-D sum** — different platform, different mechanism, no compile error.
- A bool-input variant of the same 3-D shape surfaces downstream as a subprocess crash (`SubprocException`) on fmax/maximum/minimum rows; possibly the same lowering path — worth triaging together.

**Comment by gavinwang269** [comment]
### Maintainer note (cross-reference, filed ~7h after this issue)

While extending this repro to the downstream inductor(TE) opinfo family that motivated this report, a further deterministic finding superseded the "leading-axis stride swap" reading of the same numbers:

- The inductor broadcast-backward failures (35 f16/f32 rows, incl. plain `aten.add` grads) reproduce as a **first-launch program-loss**: on the first launch of a compiled kernel, all programs with `pid >= group_tail` silently do not land their stores (deterministic stale cells); the same binary relaunched in-process is complete. The observed "wrong values equal a sum along the other axis" in multi-program grids is the stale/unwritten buffer, not a reduction-lowering swap.
- Full minimal repro + exclusion chain (binary-follows-not-launcher, metadata field-for-field identical, shape sensitivity) filed as **#2292**, which also notes a `SubprocException` / `ERR99999` crash observation possibly in the same launch-path family.

The single-program `k3d` form in this issue (grid=(1,), 50/50 wrong, `torch.zeros` output) is the one case where "unwritten buffer" may not fully explain the observed values, so please triage **#2291 + #2292 together**: if the first-launch loss explains both, this issue can be closed as a duplicate; if the single-program miscompile stands independently, keep it open — the repro here is bit-identical across fresh processes and we can re-validate either reading on request.

**Comment by hongziqi** [comment]
Thanks for your feedback. We will arrange relevant engineers to investigate this issue later.
